Improve graceful shutdown

turing85 opened the pull request:
- (Re-)Introduce Shutdown Readiness check
- Always tear down Shutdown Readiness check during shutdown
- Add optional delay between pre-shutdown phase and shutdown phase to give the infrastructure time to stop routing traffic to the application

geoand commented:
I would really much prefer if the changes here split up into two (or more?) commits :

 - the actual change
 - the cosmetic improvements attempted by the PR

geoand commented on extensions/smallrye-health/deployment/src/test/java/io/quarkus/smallrye/health/test/BlockingNonBlockingTest.java:
```diff
@@ -1,8 +1,7 @@
 import static io.restassured.RestAssured.given;
-import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.*;
```
No star imports please

turing85 replied:
Done.

geoand commented on extensions/smallrye-health/deployment/src/test/java/io/quarkus/smallrye/health/test/HealthCheckDefaultScopeTest.java:
```diff
@@ -5,12 +5,11 @@
-import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.Matchers.containsInAnyOrder;
-import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.*;
```
No star imports please

turing85 replied:
Done.

geoand commented on extensions/smallrye-health/deployment/src/test/java/io/quarkus/smallrye/health/test/HealthCheckOnManagementInterfaceDisabledTest.java:
```diff
@@ -1,9 +1,9 @@
-import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.Matchers.containsInAnyOrder;
-import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.*;
```
And here

turing85 replied:
Done.

Ladicek commented on core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownRecorder.java:
```diff
@@ -16,41 +15,36 @@ public class ShutdownRecorder {
             preShutdown.await();
+
+            if (shutdownConfig.isDelaySet()) {
+                Thread.sleep(shutdownConfig.delay.get().toMillis());
```
Maybe interruption here should be interpreted as "exit pre-shutdown, go to shutdown immediately"?

turing85 replied:
Sure can do. Should we also re-interrupt the current thread (`Thread.currentThread().interrupt()`)?

turing85 replied:
Done.

Ladicek commented on core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownConfig.java:
```diff
@@ -18,8 +18,22 @@ public class ShutdownConfig {
     @ConfigItem
     public Optional<Duration> timeout;
 
+    /**
+     * A delay between the pre-shutdown phase and the actual shutdown phase.
+     *
+     * If, for example, smallrye-health is loaded, then the readiness probe will be torn down, the delay will be waited
+     * (with the server still being up) and only then the shutdown is executed. This gives the infrastructure time to detect
+     * that the application is shutting down and stop routing traffic to the instance.
+     */
```
```suggestion
    /**
     * Delay between shutdown being requested and actually initiated. Also called the pre-shutdown phase.
     * In pre-shutdown, the server continues working as usual, except a readiness probe starts reporting "down"
     * (if the {@code smallrye-health} extension is present). This gives the infrastructure time to detect
     * that the application instance is shutting down and stop routing traffic to it.
     */
```

turing85 replied:
Done

Ladicek commented on extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java:
```diff
@@ -355,6 +356,14 @@ public void kubernetes(NonApplicationRootPathBuildItem nonApplicationRootPathBui
 
+    @BuildStep
+    AdditionalBeanBuildItem shutdownHealthCheck() {
+        return AdditionalBeanBuildItem.builder()
```
I don't know if it's possible (should be, this build step should be able to consume `ShutdownConfig`), but maybe we could do this only if a delay is configured? If it isn't, it makes no sense to have this health check, does it?

turing85 replied:
@Ladicek exactly. The problem is that this should be a runtime configuration. The delay highly depends on the infrastructure. For example, if we configure k8s/openshift to check readiness every 10 seconds, and we configure that 3 consecutive fails of the check to constitute "un-readiness", we would need to configure a timeout of at least 40 seconds (3 * 10 for the three checks, and an additional 10 if we "just missed" the previous check).

If we have a way to create a bean only if some runtime-constaints are met, I would be glad to include the bean condietionally.

Ladicek replied:
Oh, it's a runtime config. My bad, I assumed it's a build-time config.

Ladicek replied:
And no, there's no way to include / exclude a bean based on runtime conditions.

Ladicek commented:
Nice work!

turing85 commented:
> I would really much prefer if the changes here split up into two (or more?) commits :
> 
> * the actual change
> * the cosmetic improvements attempted by the PR

Multiple commits, or multiple PRs?

geoand commented:
Multiple commits in the same PR would be fine

turing85 commented:
> Multiple commits in the same PR would be fine

On it. But I won't split the changes on the tests. I do not feel it is worth the effort.

turing85 committed:
Improve graceful shutdown:
- Add compile-time property "quarkus.shutdown.delay-enabled" with a default of "false" to control whether the delay should be applied and the Shutdown Readiness check should be loaded
- Add optional delay between pre-shutdown phase and shutdown phase to give the infrastructure time to stop routing traffic to the application
- Tear down Shutdown Readiness check during shutdown if check is present

turing85 committed:
Refactoring: unify if-statements

turing85 committed:
Refactoring: extract methods

turing85 committed:
Refactoring: use stream's toList() method

turing85 committed:
Refactoring: extract constant for config key

turing85 commented:
@geoand done. I have extracted the refactorings into separate commits. Care to take another look?

turing85 commented:
@Ladicek I have incorporated your review comments. Care to take another look?

turing85 commented:
*sigh* look like all health tests have to be modified, right?

turing85 commented:
I added a compile-time property `quarkus.smallrye-health.add-shtudown-readiness-check`, which defaults to `false` in order to control whether the `ShutdownReadinessCheck` is added. Consequently, I re-worked the tests. I ran some test suites that failed before, and they succeeded (including the `tcks`). Let's see what the pipeline says.

geoand commented:
I am worried about this as we definitely do not want to introduce any breaking changes because of this feature, which means that all existing tests should remain as they are.

geoand commented:
So it seems like you were able to make the tests pass without any changes, that's great.

I have an alternative proposal to `quarkus.smallrye-health.add-shutdown-readiness-check`. 
Why not make `quarkus.smallrye-health.delay` a build time property and use this as the control for whether the `ShutdownReadinessCheck` is added? 

The reason is that with the current proposal the user needs to configure two properties to get the desired behavior, which is very counterintuitive.

turing85 commented:
> So it seems like you were able to make the tests pass without any changes, that's great.
> 
> I have an alternative proposal to `quarkus.smallrye-health.add-shutdown-readiness-check`. Why not make `quarkus.smallrye-health.delay` a build time property and use this as the control for whether the `ShutdownReadinessCheck` is added?
> 
> The reason is that with the current proposal the user needs to configure two properties to get the desired behavior, which is very counterintuitive.

@geoand see https://github.com/quarkusio/quarkus/pull/39206#discussion_r1514020232

In short: I do not think that it is a good user experience when the user has to re-compile the application, just becase we - for example - change the k8s health check settings.

geoand commented:
I think it's a far worse experience to have to configure two different properties to get the desired behavior

turing85 commented:
@geoand Additionally, I do not think we can move the `delay` property into `quarkus.smallrye-health`, but have to keep it as `quarkus.shutdown.delay` since the delay is (also) evaluated in the core, and `smallyre-health` might not be loaded.

geoand commented:
As it is a new property is no reason why `quarkus.shutdown.delay` can't be a build time property.

Ladicek commented:
If `quarkus.shutdown.timeout` is a runtime property, it would be highly counter-intuitive for `quarkus.shutdown.delay` to be a build-time property. At the same time, we do need a build-time property. Alas, it would be too late to add a build-time property like `quarkus.shutdown.graceful-enabled`, and I find it very weird to add a build-time property to the SmallRye Health extension. I propose we add a build-time property `quarkus.shutdown.delay-enabled`, defaulting to `false`. If `false`, the `quarkus.shutdown.delay` config is ignored, there's no delay, and the shutdown readiness probe is not installed. If `true`, it's the opposite -- we wait for the configured time between pre-shutdown and shutdown, and we install the shutdown readiness probe.

turing85 commented:
@geoand are you okay with this?

turing85 commented:
@geoand are you okay with @Ladicek proposal?

geoand commented:
Sure, as long as the documentation is updated to reflect this - I don't expect users to have to read the javadoc of the properties to figure things out

turing85 committed:
Refactoring: reformat lifecycle.adoc to the new format (one sentence per line), include duration snippet

turing85 commented:
@geoand @Ladicek pushed. The docs are also updated.

Ladicek commented on core/deployment/src/main/java/io/quarkus/deployment/shutdown/ShutdownBuildTimeConfig.java:
```diff
@@ -0,0 +1,18 @@
+@ConfigRoot(phase = ConfigPhase.BUILD_TIME)
+public class ShutdownBuildTimeConfig {
+
+    /**
+     * Delay between shutdown being requested and actually initiated. Also called the pre-shutdown phase.
+     * In pre-shutdown, the server continues working as usual, except a readiness probe starts reporting "down"
+     * (if the {@code smallrye-health} extension is present). This gives the infrastructure time to detect
+     * that the application instance is shutting down and stop routing traffic to it.
+     */
```
> Whether Quarkus should wait between shutdown being requested and actually initiated.
> This delay gives the infrastructure time to detect that the application instance is shutting down and stop routing traffic to it.

Ladicek replied:
Or something like that.

turing85 replied:
hmm... :smile: 
https://github.com/quarkusio/quarkus/pull/39206#discussion_r1513991091

Ladicek replied:
This is the new build-time config property, `delay-enabled`.

turing85 replied:
Done

Ladicek commented:
Also the commit message of the first commit needs updating. LGTM otherwise!

turing85 commented:
> Also the commit message of the first commit needs updating. LGTM otherwise!

What exactly needs to be updated? I thought I catched everything

turing85 commented:
looks like only the MP OTel tests failed, which are unrelated :tada:

geoand commented:
Mind rebasing onto `main`? I am not convinced the test failure is unrelated

turing85 commented:
> Mind rebasing onto `main`? I am not convinced the test failure is unrelated

Done. I also ran the mp-otel locally (`cd tcks/microprofile-opentelemetry/ && mvn test`) successfully.
